strawberryfields/configuration.py

```diff
@@ -15,138 +15,229 @@
+
+    The configuration object contains API-related configuration options. This
+    function takes into consideration only pre-defined options.
+
+    If called without passing any keyword arguments, then a default
+    configuration object is created.
+
+    Keyword arguments:
+        authentication_token (str): the token to be used for user
+            authentication
+        hostname (str): the name of the host to connect to
+        use_ssl (bool): specifies if requests should be sent using SSL
+        port (int): the port to be used when connecting to the remote service
+        debug (bool): determines if the debugging mode is requested
+
+    Returns:
+        dict[str, dict[str, Union[str, bool, int]]]: the configuration
+            object
+    """
+    hostname = kwargs.get("hostname", "localhost")
+    use_ssl = kwargs.get("use_ssl", True)
+    port = kwargs.get("port", 443)
+    debug = kwargs.get("debug", False)
+
+    config = {
+        "api": {
+            "authentication_token": authentication_token,
+            "hostname": hostname,
+            "use_ssl": use_ssl,
+            "port": port,
+            "debug": debug
+            }
+    }
+    return config
+
+def look_for_config_in_file(filename="config.toml"):
+    """Looks for the first configuration file to be found at certain paths.
+
+    .. note::
+
+        The following directories are checked (in the following order):
+
+        * The current working directory
+        * The directory specified by the environment variable SF_CONF (if specified)
+        * The user configuration directory (if specified)
+
+    Keyword arguments:
+        filename (str): the configuration file to look for
+
+    Returns:
+         dict[str, dict[str, Union[str, bool, int]]] or None: the
+             configuration object that was loaded
+    """
+
+    # Search the current directory, the directory under environment
+    # variable SF_CONF, and default user config directory, in that order.
+    current_dir = os.getcwd()
+    sf_env_config_dir = os.environ.get("SF_CONF", "")
+    sf_user_config_dir = user_config_dir("strawberryfields", "Xanadu")
+
+    directories = [current_dir, sf_env_config_dir, sf_user_config_dir]
+    for directory in directories:
+        filepath = os.path.join(directory, filename)
+        try:
+            parsed_config = load_config_file(filepath)
+            break
+        except FileNotFoundError:
+            parsed_config, filepath = None, None
+
+    return parsed_config, filepath
+
+def load_config_file(filepath):
+    """Load a configuration object from a TOML formatted file.
+
+    Args:
+        filepath (str): path to the configuration file
+
+    Returns:
+         dict[str, dict[str, Union[str, bool, int]]]: the configuration
+            object that was loaded
+    """
+    with open(filepath, "r") as f:
+        config_from_file = toml.load(f)
+    return config_from_file
+
+def update_with_other_config(config, other_config):
```

Comment:
This could be a case where a more verbose name makes it slightly harder to parse. I would suggest `update_config` is sufficient here, as it is similar to the `dict.update()` (and feels quite pythonic)

Reply:
Actually, is this method needed at all? Would `config.update(other_config)` work?

Reply:
I've changed this such that there the `dict.update` method is being used :)

One thing to note is that it cannot be replaced entirely. This is because assuming that not all options are defined in `kwargs`, calling on `update` on the outer dictionary does not update the inner dictionary, but overwrites it (updating is not recursive for the nested structure):

```python
In [1]: a = {'a': {'b': 3}}                                                     

In [2]: a                                                                       
Out[2]: {'a': {'b': 3}}

In [3]: b = {'a': {'c': 5}}                                                     

In [4]: b                                                                       
Out[4]: {'a': {'c': 5}}

In [5]: a.update(b)                                                             

In [6]: a                                                                       
Out[6]: {'a': {'c': 5}}

# Restoring a

In [7]: a = {'a': {'b': 3}}                                                     

In [8]: a                                                                       
Out[8]: {'a': {'b': 3}}

In [9]: a['a'].update(b['a'])                                                   

In [10]: a                                                                      
Out[10]: {'a': {'b': 3, 'c': 5}}
```